source/Dominance/Lifting.lagda

```diff
@@ -6,74 +6,183 @@ Martin Escardo, January 2018, May 2020
 
 open import Dominance.Definition
 open import MLTT.Spartan
-
-module Dominance.Lifting
-        {𝓣 𝓚 : Universe}
-        (d : 𝓣 ̇ → 𝓚 ̇ )
-        (isd : is-dominance d)
-       where
+open import UF.Base
+open import UF.SIP
+open import UF.Univalence
+open import UF.FunExt
+open import UF.Equiv-FunExt
+open import UF.Equiv hiding (_≅_; ≅-refl)
+open import UF.UA-FunExt
+open import UF.Subsingletons-FunExt
+import UF.PairFun as PairFun
+
+module
+ Dominance.Lifting
+  {𝓣 𝓚 : Universe}
+  (𝓣-ua : is-univalent 𝓣)
+  (d : 𝓣 ̇ → 𝓚 ̇)
+  (isd : is-dominance d)
+ where
 
  D : Dominance
  D = (d , isd)
 
- L : ∀ {𝓥} (X : 𝓥 ̇ ) → 𝓣 ⁺ ⊔ 𝓚 ⊔ 𝓥 ̇
- L X = Σ P ꞉ 𝓣 ̇ , d P × (P → X)
-
- LL : ∀ {𝓥} (X : 𝓥 ̇ ) → 𝓣 ⁺ ⊔ 𝓚 ⊔ 𝓥 ̇
- LL X = L (L X)
-
- _⇀_ : ∀ {𝓥 𝓦} → 𝓥 ̇ → 𝓦 ̇ → 𝓣 ⁺ ⊔ 𝓚 ⊔ 𝓥 ⊔ 𝓦 ̇
+ module _ {𝓥} where
+  L : (X : 𝓥 ̇) → 𝓣 ⁺ ⊔ 𝓚 ⊔ 𝓥 ̇
+  L X = Σ P ꞉ 𝓣 ̇ , (P → X) × d P
+
+  is-defined : {X : 𝓥 ̇} → L X → 𝓣 ̇
+  is-defined (P , (ϕ , dP)) = P
+
+  _↓ = is-defined
+
+  ↓-is-dominant : {X : 𝓥 ̇} → (x̃ : L X) → is-dominant D (x̃ ↓)
+  ↓-is-dominant (P , (ϕ , dP)) = dP
+
+  value : {X : 𝓥 ̇} → (x̃ : L X) → x̃ ↓ → X
+  value (P , (ϕ , dP)) = ϕ
+
+
+ module _ {𝓥 : _} {X : 𝓥 ̇} where
+  open sip
+
+  fam-str : (P : 𝓣 ̇) → 𝓣 ⊔ 𝓥 ̇
+  fam-str P = P → X
+
+  fam-sns-data : SNS fam-str (𝓣 ⊔ 𝓥)
+  fam-sns-data = ι , ρ , θ
+   where
+    ι : (u v : Σ fam-str) → ⟨ u ⟩ ≃ ⟨ v ⟩ → 𝓣 ⊔ 𝓥 ̇
+    ι (P , u) (Q , v) (f , _) = u ＝ v ∘ f
+
+    ρ : (u : Σ fam-str) → ι u u (≃-refl ⟨ u ⟩)
+    ρ _ = refl
+
+    h : {P : 𝓣 ̇} {u v : fam-str P} → canonical-map ι ρ u v ∼ -id (u ＝ v)
+    h refl = refl
+
+    θ : {P : 𝓣 ̇} (u v : fam-str P) → is-equiv (canonical-map ι ρ u v)
+    θ u v = equiv-closed-under-∼ _ _ (id-is-equiv (u ＝ v)) h
+
+  fam-≅ : (u v : Σ fam-str) → 𝓣 ⊔ 𝓥 ̇
+  fam-≅ (P , u) (Q , v) =
+   Σ f ꞉ (P → Q) , is-equiv f × (u ＝ v ∘ f)
+
+  characterization-of-fam-＝ : (u v : Σ fam-str) → (u ＝ v) ≃ fam-≅ u v
+  characterization-of-fam-＝ = characterization-of-＝ 𝓣-ua fam-sns-data
+
+  _≅_ : L X → L X → 𝓣 ⊔ 𝓥 ̇
+  (P , u , dP) ≅ (Q , v , dQ) =
+    Σ f ꞉ P ⇔ Q , u ∼ v ∘ pr₁ f
+
+  ≅-refl : (u : L X) → u ≅ u
+  ≅-refl u = (id , id) , λ _ → refl
+
+  -- TODO: move or find in library
+  Σ-assoc-equiv
```

Comment:
@martinescardo Is there somewhere in the library that already has this helper function, or to where you would like me to move it?